cloudinit/sources/DataSourceVMware.py

```diff
@@ -0,0 +1,780 @@
+        self.vendordata_raw = vd
+
+        # Purge any sensitive information.
+        self._purge_sensitive_data()
+
+        # get_data returns true if there is any available metadata,
+        # userdata, or vendordata.
+        if self.metadata or self.userdata_raw or self.vendordata_raw:
+            return True
+        else:
+            return False
+
+    def setup(self, is_new_instance):
+        """setup(is_new_instance)
+
+        This is called before user-data and vendor-data have been processed.
+
+        Unless the datasource has set mode to 'local', then networking
+        per 'fallback' or per 'network_config' will have been written and
+        brought up the OS at this point.
+        """
+
+        host_info = wait_on_network(self.metadata)
+        LOG.info("got host-info: %s", host_info)
+
+        # Reflect any possible local IPv4 or IPv6 addresses in the guest
+        # info.
+        advertise_local_ip_addrs(host_info)
+
+        # Ensure the metadata gets updated with information about the
+        # host, including the network interfaces, default IP addresses,
+        # etc.
+        self.metadata = util.mergemanydict([self.metadata, host_info])
+
+        # Persist the instance data for versions of cloud-init that support
+        # doing so. This occurs here rather than in the get_data call in
+        # order to ensure that the network interfaces are up and can be
+        # persisted with the metadata.
+        self.persist_instance_data()
+
+    @property
+    def network_config(self):
+        if "network" in self.metadata:
+            LOG.debug("using metadata network config")
+        else:
+            LOG.debug("using fallback network config")
+            self.metadata["network"] = {
+                "config": self.distro.generate_fallback_config(),
+            }
+        return self.metadata["network"]["config"]
+
+    def get_instance_id(self):
+        # Pull the instance ID out of the metadata if present. Otherwise
+        # read the file /sys/class/dmi/id/product_uuid for the instance ID.
+        if self.metadata and "instance-id" in self.metadata:
+            return self.metadata["instance-id"]
+        with open(PRODUC_UUID_FILE_PATH, "r") as id_file:
+            self.metadata["instance-id"] = str(id_file.read()).rstrip().lower()
+            return self.metadata["instance-id"]
+
+    def get_public_ssh_keys(self):
+        if "public-keys-data" in self.metadata:
+            return sources.normalize_pubkey_data(
+                self.metadata["public-keys-data"]
+            )
+        if "public_keys_data" in self.metadata:
+            return sources.normalize_pubkey_data(
+                self.metadata["public_keys_data"]
+            )
+        elif "public-keys" in self.metadata:
+            return sources.normalize_pubkey_data(self.metadata["public-keys"])
+        elif "public_keys" in self.metadata:
+            return sources.normalize_pubkey_data(self.metadata["public_keys"])
```

Comment:
Worth just doing this?
```suggestion-
        for key_name in "public-keys-data", "public_keys_data", "public_keys", "public_keys":
            if key_name in self.metadata:
                return sources.normalize_pubkey_data(self.metadata[key_name])
            return sources.normalize_pubkey_data(
                self.metadata["public_keys_data"]
            )
        elif "public-keys" in self.metadata:
            return sources.normalize_pubkey_data(self.metadata["public-keys"])
        elif "public_keys" in self.metadata:
            return sources.normalize_pubkey_data(self.metadata["public_keys"])
```

Reply:
Now you are just showing off :)